### PR TITLE
Rework readme

### PR DESCRIPTION
**Pull request by pbourguignon**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -22,46 +22,96 @@ A pure Python micro-framework supporting seamless lazy and concurrent evaluation
...
 
 
-In essence, the package allows to write *functional* code directly in Python: statements merely specify relationships among *variables* through *operations*.
-Evaluation of any variable given the values of other variables is then de facto:
+Introduction
+''''''''''''
 
-- **lazy**: only operations participating in the determination of the requested value are executed,
-- **concurrent**: operations can be executed by a thread pool of arbitrary size.
+Paragraph adds the *functional programming paradigm* to Python in a minimal and almost seamless fashion. One additional class, ``Variable``, and a
```
Comment (trendelkampschroer) [inline README.rst]: Superfluous 'almost seamless'

File: README.rst [hunk 2]
```diff
@@ -22,46 +22,96 @@ A pure Python micro-framework supporting seamless lazy and concurrent evaluation
...
-A glossary is provided below, which should clarify most concepts implemented in this module. Note that the usage of some terms may slightly differ from
-their standard definitions, reading it through before diving into the code is therefore highly recommended irrespective of the reader's familiarity with the
-topic.
 
+The few lines above fully instantiate a computation graph, here in its simplest form with just one equation relating ``x``, ``y`` and ``s`` via the function
+``add``. Given values for the input variables ``x`` and ``y``, the value of ``s`` is resolved as follows:
 
-Getting started
-===============
+>>> pg.evaluate([s], {x: 5, y: 10})
```
Comment (trendelkampschroer) [inline README.rst:42]: It should be possible to just evaluate `s` for the sake of convenience, but I understand if this feature cannot be shipped right away.

File: README.rst [hunk 3]
```diff
@@ -22,46 +22,96 @@ A pure Python micro-framework supporting seamless lazy and concurrent evaluation
...
-=============
+
+Eager currying
+  Invoking an op with invariable arguments (that is, arguments that are not of type ``Variable``) just returns an invariable value: evaluation is
+  eager whenever possible. If only a subset of inputs is a variable, say ``x_p``, the computation graph can be simplified using ``solve``, which returns a new
+  variable:
+  
+  >>> x_p = Variable("x_p")
+  >>> u = add.op(s, t)
+  >>> u_xp = pg.solve([u], {x: x_p, y: 10, z: 50})[0]
```
Comment (trendelkampschroer) [inline README.rst]: Consider renaming `x_p -> v` and `u_xp -> u_v`. Solving the equation yields `x` as a function of `v`, `x_v`!

File: README.rst [hunk 4]
```diff
@@ -22,46 +22,96 @@ A pure Python micro-framework supporting seamless lazy and concurrent evaluation
...
+Paragraph adds the *functional programming paradigm* to Python in a minimal and almost seamless fashion. One additional class, ``Variable``, and a
+function decorator, ``op``, is all it takes to turn regular Python code into a *computation graph*, i.e. a computer representation of a system of
+equations:
 
-In addition, relationships among variables can be traversed in both directions, allowing a form of backpropagation of
-information through the computation network that would be cumbersome to implement in an imperative manner.
+>>> import paragraph as pg
+>>> import operator
+>>> x, y = pg.Variable("x"), pg.Variable("y")
+>>> add = pg.op(operator.add())
```
Comment (ronaldrichter) [inline README.rst]: Is it correct to call `operator.add` here?

File: README.rst [hunk 5]
```diff
@@ -22,46 +22,96 @@ A pure Python micro-framework supporting seamless lazy and concurrent evaluation
...
+  variable:
+  
+  >>> x_p = Variable("x_p")
+  >>> u = add.op(s, t)
+  >>> u_xp = pg.solve([u], {x: x_p, y: 10, z: 50})[0]
+  
+  Here, ``u_xp`` is a different variable from ``u``: it depends on a single input variable (``x_p``), and it knows nothing about a variable ``y`` or ``z``,
+  instead storing a reference to the value of their sum ``t``, i.e. ``60``.
+
+  Thus, ``pg.session.solve`` acts much as ``functools.partial``, except it simplifies the system of equations where possible by executing dependent
```
Comment (trendelkampschroer) [inline README.rst:76]: Superfluous 'dependent'

**Review by ronaldrichter — APPROVED** [review]

**Review by trendelkampschroer — APPROVED** [review]

File: README.rst (hunk 4 above)
Comment (pbourguignon) [reply README.rst]: Oops, good point!

File: README.rst (hunk 2 above)
Comment (pbourguignon) [reply README.rst:42]: I'm no big fan of the "one-or-a-list-of" thingy, sorry...

File: README.rst (hunk 1 above)
Comment (pbourguignon) [reply README.rst]: Done

File: README.rst (hunk 3 above)
Comment (pbourguignon) [reply README.rst]: No, it is yielding ``u`` as a function of ``x_p`` (``_p`` is for "prime").

Comment (trendelkampschroer) [reply README.rst]: Yes sorry, I mean `u` as a function of `v`, otherwise you'd end up with a double subscript u_{x_{p}} which is not readable.

Comment (pbourguignon) [reply README.rst]: But I should improve that, true. There are two separate messages to be conveyed here: 

1. If some input variables are given a value, but not all, ``solve`` returns a simplified computation graph. For this, ``x_p`` is just useless.
2. If you pass another variable (which may itself depend on anything) as the argument value for the input variable ``x``, ``solve`` returns a new computation graph with the resulting dependencies pieced together. Let me try to make this clearer.

File: README.rst (hunk 5 above)
Comment (pbourguignon) [reply README.rst:76]: Not really superfluous, it is really executing _only_ dependent variables.

File: README.rst (hunk 3 above)
Comment (trendelkampschroer) [reply README.rst]: Sorry `u_v = solve(u, {x: v, y: 10, z: 50})`  is more readable than the list version `u_v = solve([u], {...})[0]`. At some point the 'total number of brackets tolerance' of the average user gets exhausted...

Comment (pbourguignon) [reply README.rst]: So, I simplified the above example, simply removing the definition of x_p and the argument for x in solve.
I'll add a subsection on "Graph composition" to explain point 2 as a separate functionality (which it is).

Comment (pbourguignon) [reply README.rst]: As for the list thingy, it sure is more readable without the additional brackets. But this is not the only use case, and it is paramount to be able to pass in a list (and I'd expect that to be the most frequent use case, by far).

**Review by mathiaseitz — APPROVED** [review]

**Pull request merged**